### PR TITLE
Sequencer module for execution of timed thruster sequences

### PR DESCRIPTION
**Pull request by kodsurf**
So .... 

I would like to finalize current design iteration with thruster sequencer module by opening this PR. 

All Thruster sequences that were hardcoded  - work as intended without causing hard fault loop. 

It is not supposed that multiple thruster sequences would run at the same time - nevertheless most times sequences are successfully running in parallel. 

Therefore I submit already working solution for running thruster tests and firing sequences. 

Print messages are there for a purpose of logging and validation. In case we need to run another radiation test, this time including thruster board - logging messages (printed to debug uart) should document every action, including bytes that thruster is replying. This way with the log it would be able to validate that OBC is sending correct requests, and tests/operations are running as intended.

Example of logging output
image
image
image

Note : 

https://github.com/RobertK66/obc_1769_core/issues/49 and https://github.com/RobertK66/obc_1769_core/issues/52 are still open and to be resolved during further design iterations. 

As I mentioned before - it is not allowed to run parallel thruster sequences ( different tests cannot run at the same time !!!!). But from SW point of view - it should be possible without hard fault.  This is what needs to be achieved by resolving #49 
(hard fault loop only occures with thruster board turned on/ when thruster sends reply with too high frequency >1Hz) 

Currently frequency between SET requests is hardcoded to be 0.5Hz (which causes no problems). But the reasons for such a behaviour is yet unclear. @RobertK66  Need your help with that.

### CODE REVIEW

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c]: this is a good idea. But has to be implemented by avoiding argv ...

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:429]: while( next_register_index<108 )

next_register_index should be less then MAXIMUM REGISTER INDEX that is avaliable based on amount of bytes that we requested from thruster. 

In case of our hardcoded ReadAllRegisters() request  - the amount of bytes that we ask should cover 108 registers, there fore exit condition will reach. 

BUT ! If we make another custom request for more then 4 byte but less then enough to cover all 108 registers !! next_register_index will be less then 108. And this will cause infinite loop 

I should do function that calculates max_register_index that is reached by length of payload. To make parser more universal.

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:440]: But still - this exit condition is a MUST !

If for some reason thruster will reply less bytes. Or some bytes arre corrupted or lost. To avoid stuck into infinite loop we add break condition based on maximum allowed proccessing time. 

It should be less then 100ms. 


I will probably formulate another issue regarding parser. But for next design iterations.

For now it should be fine

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:479]: Here is no problem. It is solved.Here is just my dirty code .....

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:483]: I want to leave debug messages commented instead of deleting it (in case I ever need to investigate into parser again)

For now parser works as expected.

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:529]: Not  THR_EXECUTION_INDEX

but 
THR_HARDCODED_SEQUENCES[procedure_id].execution_index

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:770]: Not only for Hot Standby 

For all sequences that require to wait and monitor. 

I will need to update this function WHEN CLIENT (ENPULSION) or CLIMB system engineers will provide more details on what have to be monitored (and what to do when things go wrong, when to interrupt sequence)

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c]: I would have to redesign all GeneralRequest functions to avoid this mess with sprintf() and argv ...

This is total stupid.  this argv approach was inherited from commands .... but totally redundent here. 

When I realized that this is stupid - it was too late and whole sequencer module was almost complete .... This is why I decided to "follow the style" to produce solution that works as fast as possible without argv redesign

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:885]: This is just function that does nothing in a sequence.  I thought that ..... it might be handy someday :) But for now it is redundent

My logic was that there could be something like jump_sequence(uint8_t index) 

which jumps to a specific index at hardcoded sequence. Can be handy if we need to implement some loops within sequences. 

For example - while waiting for right conditions we loop through monitoring section of sequence.  

to develop this further and implement - need details from SYSTEM ENGINEERING.

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c]: Yes, this stupidity again. 

but no other way if needed to make it work with argv ....

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:923]: Functions like thr_ramp() and wait_and_monitor() - need to have a concept of "substage" 

substages are implemented via switch (case)

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c]: This is because we cant send SET/READ requests and instantly expect reply !
We have to introduce WAIT stages, without delaying the main loop. 

Therefore substages are implemented

File: src/mod/l4_thruster.c
Comment (kodsurf) [inline src/mod/l4_thruster.c:947]: If we waited long enough - substage index is increased and thr_ramp() function will be executed during next iteration with increased substage index

File: src/mod/l2_debug_com.c
Comment (RobertK66) [inline src/mod/l2_debug_com.c:23]: keep it as is at this moment. But I plan to move this 'debug feature' of current running module somewhere to the base / mainloop code... (see #17 )

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:48]: As I did it till now, I always implemented the argc/argv Command line versions somwhere in L7. This is to have the future possibility at some stage of the project to get rid of the string to number conversions in the final flight version. For all commands we need as internal API or as groundstation command there should be the 'binary version' having its parameters as named method pars.

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:58]: Pls. remove this MRAM 'experimental'/#play around code prior to margin this to develop, thx.

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:65]: Ah ok I see, you use this argc/argv as a internal 'step callback, .....
... let me think about that till end of review ;-)! ....

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:80]: is this a counter starting at 0 or a pointer also!? From name I asume its a counter.

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:72]: is this a pointer to (current) thr_sequences_t !? Then you could/should declare it so:

`thr_sequences_t * execution_index;`

or better :-):
`thr_sequences_t * current_step_ptr;`

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:153]: I think for the OBC C code we should have this only in enums -> nicley readable in code (and I think in debugger). For the logger there can be some means of translating the C-coded enums to texts (outside of the OBC software!) as I explained when arguing for the binary event API to be used.....

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:163]: this is the classical enum to be used here. You can read it in code and it uses the numbers when compiled.

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:88]: <--- This is comment to line 97. (Can not attach direct because was not changed).
Not sure I understand what you are doing/trying to do here ....
Maybe I get it when I read the code later ....

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:238]: Comment to line 228/229:   This is a verry dangerous implementation here !!!
Imagine that you get 2 different Requests called from outside (e.g. one from debug CLI or running sequence and one from groundstation radio RX only to use some scenario you can run into)! The 2nd one will make its TX to the Thr UART and overwrite the expected ReceiveBuffer count, I think. Yiou have to have some 'queing' or at least blocking in place to avoid such scenarios !!!

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:268]: This is pure 'development debugging code'. Should not be in final flight version at all. Either get rid of it or try to 'move it up' somehow into L7. (E.g. by introducing a pure binary 'register changed' event or something simmilar....)

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:276]: PLS DO NOT USE THIS any more. I showed you how to use String/raw events in an orderly manner during our last session. And I will provide this also with #17 ...

File: src/mod/l4_thruster.c
Comment (RobertK66) [inline src/mod/l4_thruster.c:304]: replace with simple event 'Thr_rx_OK' - can be a 2 byte event with level info, module ID of thruster and only event ID (known and defined by Thruster module. -> this 2 bytes can be translated to this text by debug tool outside of OBC then!

Comment (RobertK66) [reply src/mod/l4_thruster.c:304]: I will not comment this in the rest of code but opened #57

File: src/mod/l4_thruster.c
Comment (RobertK66) [reply src/mod/l4_thruster.c:440]: ok, adding timeout has to be tackled for this Thruster communications. So go on and formulate the issue for that as stated ;-).

File: src/mod/l4_thruster.c
Comment (RobertK66) [reply src/mod/l4_thruster.c:483]: No problem with old code in  comments for debugging (but sometimes thay get to old to be executed (by only commenting them in) Then that's the time to delete - or rewrite better debugging code then.....

File: src/mod/modules_globals.h
Comment (RobertK66) [inline src/mod/modules_globals.h:14]: I will try to place this with #17

File: src/mod/thr/thr.c
Comment (RobertK66) [inline src/mod/thr/thr.c:198]: I was never able to reach this code without changing the line 185 to have a '<=' iso '<' !!??
Opened #58 for this.

**Review by RobertK66 — CHANGES_REQUESTED** [review]
is this version (commit d06a41d2fba3cb4b42d06dd4902412bd9e1fd4be ) still the one I should review!?
it's the lates one I can find and I also use this to debug and test your code now.

I am going to open separate issues for bugs I find and use this here as general Review & remarks.

- Having the sequences hardcoded in Flash ROM is ok for now. Somewhen we should think about storing and using them from MRAM somehow.....
- Other than the removal of deb_print_pure_debug() I do not see any blockers to merge this into develop. 
- Either you make the change to STR event and go on with this or make #57 work before merging. I let you decide on that.

File: src/mod/l4_thruster.c
Comment (kodsurf) [reply src/mod/l4_thruster.c:80]: I call a function that is being executed in a sequence - "stage"

There are functions like ramp / wait_and monitor  which have stages of their own (implemented with switch-case) 

substage_index is used to operate this kind of functions

Comment (kodsurf) [reply src/mod/l4_thruster.c:80]: if it was OOP  I would make this variable private. Not accessable by user. 

This is a "service" variable

**Comment by RobertK66** [comment]
Will be replaced with new version (from feature/thruster_experimental) ....

**Pull request closed without merging**
